src/Language/Java/Inline.hs

```diff
@@ -399,14 +399,29 @@ blockQQ input = case Java.parser Java.block input of
       -- lifted, we'll support returning unboxed values, just like `call` does.
       castReturnType funcall = [| unsafeUncoerce . coerce <$> $funcall |]
 
+-- | Non capture-avoiding substitution. The argument type should not contain any
+-- variable binding forms.
+substType :: [(TH.Name, TH.Type)] -> TH.Type -> TH.Type
+substType ctx = everywhere (mkT go)
+  where
+    go (TH.VarT name) | Just ty' <- lookup name ctx = ty'
+    go TH.ForallT{} = error "substType: forall not supported."
+    go ty = ty
+
 -- Recursively unfold type synonyms, if any.
 unfoldTypeTySyn :: TH.Type -> Q TH.Type
-unfoldTypeTySyn = everywhereM (mkM go)
+unfoldTypeTySyn ty
+  | (TH.ConT name, actuals) <- decomp [] ty =
+    TH.reify name >>= \case
+      TH.TyConI (TH.TySynD _ (map fromBndr -> parms) ty') -> do
+        when (length actuals < length parms) $
+          -- In principle GHC would never let this happen anyways.
+          fail "Internal error: type synonym not fully saturated."
```

Comment:
Using `error` instead of fail would give the location of the error call. Alternatively, augment the message to explain which module or package it came from.